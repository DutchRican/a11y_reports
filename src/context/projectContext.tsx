import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { createContext, ReactNode, useContext, useMemo, useState } from 'react';
import { toast } from 'react-toastify';
import { deleteProject, fetchProjects } from '../api/projects';
import { Project } from '../types';

type ProjectContextType = {
	projectID: string | undefined;
	setProjectID: (id: string | undefined) => void;
	availableProjects: Project[];
	isLoadingProjects: boolean;
	isRefetchingProjects: boolean;
	currentProject: Project | undefined;
	projectsError: Error | null;
	removeProject: (projectId: string) => void;
};

const ProjectContext = createContext<ProjectContextType | undefined>(undefined);

export const ProjectProvider = ({ children }: { children: ReactNode }) => {
	const [projectID, setProjectID] = useState<string | undefined>(undefined);
	const queryClient = useQueryClient();

	const { data: availableProjects = [], isLoading: isLoadingProjects, isRefetching: isRefetchingProjects, error: projectsError } = useQuery<Project[], Error>({
		queryKey: ['projects'],
		queryFn: fetchProjects,
	});

	const { mutate: removeProject } = useMutation({
		mutationFn: deleteProject,
		onSuccess: () => {
			queryClient.invalidateQueries({ queryKey: ['projects'] });
			toast.success('Project deleted successfully');
		},
		// Assuming deleteProject is defined in your API module
		onMutate: async (projectId: string) => {
			deleteProject(projectId);
		},
		onError: (error: any) => {
			toast.error(`Failed to delete project: ${error.message}`);
		},
	});

	const currentProject = availableProjects.find(project => project._id === projectID);

	const contextValue: ProjectContextType = useMemo(() => ({
		projectID,
		setProjectID,
		availableProjects,
		isLoadingProjects,
		isRefetchingProjects,
		currentProject,
		projectsError,
		removeProject,
	}), [projectID, availableProjects, isLoadingProjects, isRefetchingProjects, currentProject, projectsError, removeProject]);

	return (
		<ProjectContext.Provider value={contextValue}>
			{children}
		</ProjectContext.Provider>
	);
};

// eslint-disable-next-line react-refresh/only-export-components
export const useProjectContext = () => {
	const context = useContext(ProjectContext);
	if (!context) {
		throw new Error('useProjectContext must be used within a ProjectProvider');
	}
	return context;
};