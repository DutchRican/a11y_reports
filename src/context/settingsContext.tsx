import { createContext, ReactNode, useContext, useEffect, useState } from "react";

type SettingsContextType = {
	theme: string;
	setTheme: (theme: string) => void;
	isAdminMode: boolean;
	enableAdminMode: (password: string) => void;
	disableAdminMode: () => void;
	earliestFetchDate: number;
	setEarliestFetchDate: (days: number) => void;
	password: string;
};

const SettingsContext = createContext<SettingsContextType | undefined>(undefined);

export const SettingsProvider = ({ children }: { children: ReactNode }) => {
	const [theme, setTheme] = useState<string>(localStorage.getItem("theme") || "light");
	const [isAdminMode, setIsAdminMode] = useState<boolean>(false);
	const [earliestFetchDate, setEarliestFetchDate] = useState<number>(
		localStorage.getItem("earliestFetchDate") ? parseInt(localStorage.getItem("earliestFetchDate")!, 10) : 21
	);
	const [adminTimeout, setAdminTimeout] = useState<number | undefined>(undefined);
	const [password, setPassword] = useState<string>("");

	useEffect(() => {
		return () => {
			if (adminTimeout) {
				clearTimeout(adminTimeout);
			}
		};
	}, [adminTimeout]);

	const enableAdminMode = (pwd: string) => {
		if (adminTimeout) {
			clearTimeout(adminTimeout);
		}
		setPassword(pwd);
		setIsAdminMode(true);
		const timeoutId = setTimeout(() => {
			setIsAdminMode(false);
			setPassword("");
		}, 2 * 60 * 1000); // 2 minutes
		setAdminTimeout(timeoutId as unknown as number);
	};

	const disableAdminMode = () => {
		setIsAdminMode(false);
		if (adminTimeout) {
			clearTimeout(adminTimeout);
		}
	};

	const contextValue: SettingsContextType = {
		theme,
		setTheme,
		earliestFetchDate,
		setEarliestFetchDate,
		isAdminMode,
		enableAdminMode,
		disableAdminMode,
		password
	};

	return (
		<SettingsContext.Provider value={contextValue}>
			{children}
		</SettingsContext.Provider>
	);
};

export const useSettings = () => {
	const context = useContext(SettingsContext);
	if (context === undefined) {
		throw new Error("useSettings must be used within a SettingsProvider");
	}
	return context;
};
