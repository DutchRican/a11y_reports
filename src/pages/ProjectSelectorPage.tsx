import { useNavigate } from "react-router-dom";
import { useProjectContext } from "../context/projectContext";

const ProjectSelectorPage: React.FC = () => {
	const { availableProjects, isLoadingProjects, setProjectID, projectID, isRefetchingProjects } = useProjectContext();
	const navigate = useNavigate();
	const isBusy = isLoadingProjects || isRefetchingProjects;

	const handleProjectSelect = (projectId: string) => {
		if (!projectID || projectID !== projectId) {
			setProjectID(projectId);
		} else {
			navigate(`/project/${projectId}`);
		}
	};
	return (
		<div className="max-w-7/8 mx-auto mt-10 p-6 bg-white rounded-lg shadow">
			<h1 className="text-center text-2xl font-semibold mb-6">Select a Project</h1>
			<ul className="list-none p-0">
				{isBusy && (
					<li className="p-3 mb-3 bg-gray-100 rounded">Loading projects...</li>
				)}
				{availableProjects.length === 0 && !isBusy && (
					<li className="p-3 mb-3 bg-gray-100 rounded">
						No projects available. Please create a project first.
					</li>
				)}
				{!isBusy && availableProjects.map((project) => (
					<li
						key={project._id}
						onClick={() => handleProjectSelect(project._id)}
						className="p-3 mb-3 bg-gray-100 rounded cursor-pointer transition-colors hover:bg-indigo-100 flex "
					>
						<p className="font-semibold flex-1 ml-2">{project.name}</p>
						<p
							className="text-gray-500 ml-6 overflow-hidden truncate flex-3"
							title={project.description ? project.description : undefined}
						>
							{project.description || "No description available"}
						</p>

					</li>
				))}
			</ul>
		</div>
	);
};

export default ProjectSelectorPage;
