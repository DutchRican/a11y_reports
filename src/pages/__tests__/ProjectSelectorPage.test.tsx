// import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import { SettingsProvider } from '../../context/settingsContext';
import ProjectSelectorPage from '../ProjectSelectorPage';

// Mock useNavigate
const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
	useNavigate: () => mockNavigate,
}));

// Helper to mock useProjectContext
const mockSetProjectID = jest.fn();
let mockContextValue: any = {};
jest.mock('../../context/projectContext', () => ({
	useProjectContext: () => mockContextValue,
	useSettings: () => { },
}));

const Wrapper = () => (
	<SettingsProvider>
		<ProjectSelectorPage />
	</SettingsProvider>
);

describe('ProjectSelectorPage', () => {
	beforeEach(() => {
		jest.clearAllMocks();
		mockContextValue = {};
	});

	it('shows loading state', () => {
		mockContextValue = {
			availableProjects: [],
			isLoadingProjects: true,
			isRefetchingProjects: false,
			setProjectID: mockSetProjectID,
			projectID: null,
		};
		render(<Wrapper />);
		expect(screen.getByText(/Loading projects/i)).toBeInTheDocument();
	});

	it('shows no projects message', () => {
		mockContextValue = {
			availableProjects: [],
			isLoadingProjects: false,
			isRefetchingProjects: false,
			setProjectID: mockSetProjectID,
			projectID: null,
		};
		render(<Wrapper />);
		expect(screen.getByText(/No projects available/i)).toBeInTheDocument();
	});

	it('shows available projects', () => {
		mockContextValue = {
			availableProjects: [
				{ _id: '1', name: 'Project One', description: 'Desc One' },
				{ _id: '2', name: 'Project Two', description: '' },
			],
			isLoadingProjects: false,
			isRefetchingProjects: false,
			setProjectID: mockSetProjectID,
			projectID: null,
		};
		render(<Wrapper />);
		expect(screen.getByText('Project One')).toBeInTheDocument();
		expect(screen.getByText('Project Two')).toBeInTheDocument();
		expect(screen.getByText('Desc One')).toBeInTheDocument();
		expect(screen.getByText('No description available')).toBeInTheDocument();
	});

	it('calls setProjectID when selecting a new project', () => {
		mockContextValue = {
			availableProjects: [{ _id: '1', name: 'Project One', description: '' }],
			isLoadingProjects: false,
			isRefetchingProjects: false,
			setProjectID: mockSetProjectID,
			projectID: null,
		};
		render(<Wrapper />);
		fireEvent.click(screen.getByText('Project One'));
		expect(mockSetProjectID).toHaveBeenCalledWith('1');
		expect(mockNavigate).toHaveBeenCalledWith('/project/1');
	});

	it('navigates when selecting the same project', () => {
		mockContextValue = {
			availableProjects: [{ _id: '1', name: 'Project One', description: '' }],
			isLoadingProjects: false,
			isRefetchingProjects: false,
			setProjectID: mockSetProjectID,
			projectID: '1',
		};
		render(<Wrapper />);
		fireEvent.click(screen.getByText('Project One'));
		expect(mockSetProjectID).not.toHaveBeenCalled();
		expect(mockNavigate).toHaveBeenCalledWith('/project/1');
	});

	it('renders the page title', () => {
		mockContextValue = {
			availableProjects: [],
			isLoadingProjects: false,
			isRefetchingProjects: false,
			setProjectID: mockSetProjectID,
			projectID: null,
		};
		render(<Wrapper />);
		expect(screen.getByText('Select a Project')).toBeInTheDocument();
	});

	it('handles clicking multiple projects', () => {
		mockContextValue = {
			availableProjects: [
				{ _id: '1', name: 'Project One', description: 'Desc One' },
				{ _id: '2', name: 'Project Two', description: 'Desc Two' },
			],
			isLoadingProjects: false,
			isRefetchingProjects: false,
			setProjectID: mockSetProjectID,
			projectID: null,
		};
		render(<Wrapper />);
		fireEvent.click(screen.getByText('Project One'));
		expect(mockSetProjectID).toHaveBeenCalledWith('1');
		fireEvent.click(screen.getByText('Project Two'));
		expect(mockSetProjectID).toHaveBeenCalledWith('2');
	});

	it('shows "No description available" if description is undefined', () => {
		mockContextValue = {
			availableProjects: [
				{ _id: '3', name: 'Project Three' },
			],
			isLoadingProjects: false,
			isRefetchingProjects: false,
			setProjectID: mockSetProjectID,
			projectID: null,
		};
		render(<Wrapper />);
		expect(screen.getByText('No description available')).toBeInTheDocument();
	});

	it('matches snapshot for a typical project list', () => {
		mockContextValue = {
			availableProjects: [
				{ _id: '1', name: 'Project One', description: 'Desc One' },
				{ _id: '2', name: 'Project Two', description: '' },
			],
			isLoadingProjects: false,
			isRefetchingProjects: false,
			setProjectID: mockSetProjectID,
			projectID: null,
		};
		const { container } = render(<Wrapper />);
		expect(container).toMatchSnapshot();
	});
});