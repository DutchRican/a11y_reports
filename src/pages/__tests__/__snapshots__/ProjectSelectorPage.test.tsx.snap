// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ProjectSelectorPage matches snapshot for a typical project list 1`] = `
<div>
  <div
    class="max-w-7/8 mx-auto mt-10 p-6 bg-white rounded-lg shadow"
  >
    <h1
      class="text-center text-2xl font-semibold mb-6"
    >
      Select a Project
    </h1>
    <ul
      class="list-none p-0"
    >
      <li
        class="p-3 mb-3 rounded cursor-pointer transition-colors flex items-center justify-between bg-gray-100 hover:bg-indigo-100"
      >
        <div
          class="flex-1 flex items-center"
        >
          <p
            class="font-semibold ml-2"
          >
            Project One
          </p>
          <p
            class="text-gray-500 ml-6 overflow-hidden truncate"
            title="Desc One"
          >
            Desc One
          </p>
        </div>
        <div
          class="relative"
          data-test-id="hover-menu-1"
        >
          <button
            aria-haspopup="true"
            aria-label="Project actions"
            class="p-2 rounded-full hover:bg-gray-200 focus:outline-none"
          >
            <svg
              class="h-5 w-5 text-gray-600"
              fill="currentColor"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 6a2 2 0 110-4 2 2 0 010 4zM10 12a2 2 0 110-4 2 2 0 010 4zM10 18a2 2 0 110-4 2 2 0 010 4z"
              />
            </svg>
          </button>
        </div>
      </li>
      <li
        class="p-3 mb-3 rounded cursor-pointer transition-colors flex items-center justify-between bg-gray-100 hover:bg-indigo-100"
      >
        <div
          class="flex-1 flex items-center"
        >
          <p
            class="font-semibold ml-2"
          >
            Project Two
          </p>
          <p
            class="text-gray-500 ml-6 overflow-hidden truncate"
          >
            No description available
          </p>
        </div>
        <div
          class="relative"
          data-test-id="hover-menu-2"
        >
          <button
            aria-haspopup="true"
            aria-label="Project actions"
            class="p-2 rounded-full hover:bg-gray-200 focus:outline-none"
          >
            <svg
              class="h-5 w-5 text-gray-600"
              fill="currentColor"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 6a2 2 0 110-4 2 2 0 010 4zM10 12a2 2 0 110-4 2 2 0 010 4zM10 18a2 2 0 110-4 2 2 0 010 4z"
              />
            </svg>
          </button>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
